Document new upload mechanism (#835)
mozilla/glean: docs/dev/core/internal/upload.md

# Upload mechanism

The `glean-core` Rust crate does not handle the ping upload directly.
Network stacks vastly differ between platforms, applications and operating systems.

> It is best to use the available platform capabilities to implement any network communication.

Reviewer: ```suggestion
The Glean SDK leverages the available platform capabilities to implement any network communication.
```

? I merely wanted to remove the "it is best" part :-) No strong opinion on this, if you feel strongly about keeping it!

> Glean still controls all upload and coordinates the platform side with its own internals.

Reviewer 2: ```suggestion
Glean core controls all upload and coordinates the platform side with its own internals.
```

> All wrappers implement ping uploading around a common API.

Reviewer: ```suggestion
All language bindings implement ping uploading around a common API and protocol.
```

Reviewer: Is wrapper the right word here? Do we use it elsewhere in this same context? What do you think of "language binding" or "platform-specific implementation" ?

Author: language binding!
I didn't feel comfortable with wrapper to begin with, but the only other word my brain came up with was platform, and that didn't fit either.

## Upload task API

The following diagram visualizes the communication between Glean core (the Rust crate),

> a Glean wrapper (e.g. the Kotlin or Swift implementation) and a Glean end point server.

Reviewer 2: ```suggestion
a Glean language binding (e.g. the Kotlin or Swift implementation) and a Glean end point server.
```

```mermaid

> sequenceDiagram

Reviewer: Ok, I *love* this.

Author: eeeexactly

    participant Glean core
    participant Glean wrapper
    participant Server

    Glean wrapper->>Glean core: get_upload_task()
    Glean core->>Glean wrapper: Task::Upload(task)
    Glean wrapper-->>Server: POST /submit/{task.id}
    Server-->>Glean wrapper: 200 OK
    Glean wrapper->>Glean core: upload_response(200)
    Glean wrapper->>Glean core: get_upload_task()
    Glean core->>Glean wrapper: Task::Done
```

Glean core will take care of file management, cleanup, rescheduling and throttling[^1].

> [^1] Note: At this point throttling is not implemented. Follow [Bug 1543612](https://bugzilla.mozilla.org/show_bug.cgi?id=1543612) for updates.

## Available APIs

### Rust

For direct Rust consumers the global `Glean` object provides these methods:

```rust

> /// Gets the next task for an uploader. Which can be either:

Reviewer 2: Which can be either?

Author: oh. yeah, I should have not copied the old stuff.

Author: ```suggestion
/// Gets the next task for an uploader.
```

fn get_upload_task(&self) -> PingUploadTask

/// Processes the response from an attempt to upload a ping.
fn process_ping_upload_response(&self, uuid: &str, status: UploadResult)
```

See the documentation for further usage and explanation of the additional types:

* [`get_upload_task`](../../../../docs/glean_core/struct.Glean.html#method.get_upload_task)
* [`process_ping_upload_response`](../../../../docs/glean_core/struct.Glean.html#method.process_ping_upload_response)
* [`PingUploadTask`](../../../../docs/glean_core/upload/enum.PingUploadTask.html)
* [`UploadResult`](../../../../docs/glean_core/upload/enum.UploadResult.html)

> ### FFI

Reviewer 2: Did you consider using the multi-tab UI we use elsewhere for our multilanguage support?

Author: I have not, let me try that
